sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala

```diff
@@ -835,14 +832,12 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
 
   private def modeForDSV2 = mode.getOrElse(SaveMode.Append)
 
-  private def canUseV2Source(session: SparkSession, providerClass: Class[_]): Boolean = {
-    val useV1Sources =
-      session.sessionState.conf.useV1SourceWriterList.toLowerCase(Locale.ROOT).split(",")
-    val shouldUseV1Source = providerClass.newInstance() match {
-      case d: DataSourceRegister if useV1Sources.contains(d.shortName()) => true
-      case _ => useV1Sources.contains(providerClass.getCanonicalName.toLowerCase(Locale.ROOT))
+  private def lookupV2Provider() = {
+    DataSource.lookupDataSourceV2(source, df.sparkSession.sessionState.conf) match {
+      // File source v2 write path is currently broken.
+      case Some(_: FileDataSourceV2) => None
```

Comment:
This is better than before, which use configs to disable file source v2 write: https://github.com/apache/spark/pull/25465/files#diff-9a6b543db706f1a90f790783d6930a13L1605

configs can be changed, and we shouldn't allow users to use the broken file source v2 write.

Reply:
Why not add this to `lookupDataSourceV2` instead? Doesn't this logic apply to SQL as well?

Reply:
This doesn't apply to `DataFrameReader`. We can still read file source v2 with `DataFrameReader` and I don't want to break it.

Reply:
If the file sources don't support the schema from the metastore, then the read path should not be used. While it doesn't fail while inferring a schema, the source will read using the inferred schema instead of the metastore schema and that's a bug.

For example, if I create a table with 3 columns and drop one, the data and the inferred schema will have the original 3 columns. But the metastore table will have just 2 columns. Reading with v2 would bring back a column that is deleted.

Until the file sources are fixed, they should be disabled in all cases.

Reply:
I'm talking about `spark.read.parquet(path)`. It's nothing about the metastore, and the behavior is well defined in Spark 2.x:
1. if no schema is specified, e.g. `spark.read.parquet(path)`, then the schema will be inferred.
2. if schema is specified by users, e.g. `spark.read.schema(...).parquet(path)`, then schema inference will be skipped and Spark fails at runtime if the specified schema doesn't match the actual file schema.

I think what you were talking about `CREATE TABLE ... USING parquet`. This is a different case and currently file source v2 is already disabled for it.

Reply:
I think you're right that this is not affected by the metastore problem because this is reading a path-based table and not a metastore table.